File: nil/cmd/nil/internal/wallet/deploy.go
package wallet

import (
	"errors"
	"fmt"

	"github.com/NilFoundation/nil/nil/cmd/nil/internal/common"
	"github.com/NilFoundation/nil/nil/cmd/nil/internal/config"
	"github.com/NilFoundation/nil/nil/internal/types"
	"github.com/NilFoundation/nil/nil/services/cliservice"
	"github.com/NilFoundation/nil/nil/services/cometa"
	"github.com/NilFoundation/nil/nil/services/rpc/jsonrpc"
	"github.com/spf13/cobra"
)

func DeployCommand(cfg *common.Config) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "deploy [path to file] [args...]",
		Short: "Deploy a smart contract",
		Long:  "Deploy the smart contract with the specified hex-bytecode from stdin or from file",
		Args:  cobra.ArbitraryArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			return runDeploy(cmd, args, cfg)
		},
		SilenceUsage: true,
	}

	setDeployFlags(cmd)

	return cmd
}

func setDeployFlags(cmd *cobra.Command) {
	cmd.Flags().Var(
		types.NewShardId(&params.shardId, types.BaseShardId),
		shardIdFlag,
		"Specify the shard ID to interact with",
	)

	params.salt = *types.NewUint256(0)
	cmd.Flags().Var(
		&params.salt,
		saltFlag,
		"The salt for the deploy message",
	)

	cmd.Flags().StringVar(
		&params.abiPath,
		abiFlag,
		"",
		"The path to the ABI file",
	)

	params.amount = types.Value{}
	cmd.Flags().Var(
		&params.amount,
		amountFlag,
		"The amount of default tokens to send",
	)

	params.currency = types.Value{}
	cmd.Flags().Var(
		&params.currency,
		"currency",
		"The amount of contract currency to generate. This operation cannot be performed when the \"no-wait\" flag is set",
	)

	cmd.Flags().BoolVar(
		&params.noWait,
		noWaitFlag,
		false,
		"Define whether the command should wait for the receipt",
	)

	cmd.Flags().StringVar(
		&params.compileInput,
		compileInput,
		"",
		"The path to the JSON file with the compilation input. Contract will be compiled and deployed on the blockchain and the Cometa service",
	)
}

func runDeploy(_ *cobra.Command, cmdArgs []string, cfg *common.Config) error {
	if !params.currency.IsZero() && params.noWait {
		return errors.New("the \"no-wait\" flag cannot be used with the \"currency\" flag")
	}

	service := cliservice.NewService(common.GetRpcClient(), cfg.PrivateKey)

	var cm *cometa.Client
	if len(params.compileInput) != 0 {
		cm = common.GetCometaRpcClient()
	}

	if len(params.compileInput) == 0 && len(cmdArgs) == 0 {
		return errors.New("at least one arg is required (the path to the bytecode file)")
	}

	var bytecode types.Code
	var err error
	var contractData *cometa.ContractData

	if len(params.compileInput) != 0 {
		contractData, err = cm.CompileContract(params.compileInput)
		if err != nil {
			return fmt.Errorf("failed to compile the contract: %w", err)
		}
		var calldata []byte
		if len(cmdArgs) > 0 {
			if params.abiPath == "" {
				return errors.New("The ABI file required for the constructor arguments")
			}
			calldata, err = common.ArgsToCalldata(params.abiPath, "", cmdArgs)
			if err != nil {
				return fmt.Errorf("failed to pack the constructor arguments: %w", err)
			}
		}
		bytecode = append(contractData.InitCode, calldata...) //nolint:gocritic
	} else {
		var filename string
		var args []string
		if argsCount := len(cmdArgs); argsCount > 0 {
			filename = cmdArgs[0]
			args = cmdArgs[1:]
		}

		bytecode, err = common.ReadBytecode(filename, params.abiPath, args)
		if err != nil {
			return err
		}
	}

	payload := types.BuildDeployPayload(bytecode, params.salt.Bytes32())

	msgHash, contractAddr, err := service.DeployContractViaWallet(params.shardId, cfg.Address, payload, params.amount)
	if err != nil {
		return err
	}

	var receipt *jsonrpc.RPCReceipt
	if !params.noWait {
		if receipt, err = service.WaitForReceipt(cfg.Address.ShardId(), msgHash); err != nil {
			return err
		}
	} else {
		if len(params.compileInput) != 0 {
			return errors.New("the \"no-wait\" flag cannot be used with contract compilation")
		}
	}
	if receipt == nil || !receipt.AllSuccess() {
		return errors.New("deploy message processing failed")
	}

	if len(params.compileInput) != 0 {
		if err = cm.RegisterContract(contractData, contractAddr); err != nil {
			return fmt.Errorf("failed to register the contract: %w", err)
		}
	}

	if !config.Quiet {
		fmt.Print("Message hash: ")
	}
	fmt.Printf("0x%x\n", msgHash)

	if !config.Quiet {
		fmt.Print("Contract address: ")
	}
	fmt.Printf("0x%x\n", contractAddr)
	return nil
}

File: clijs/src/commands/wallet/deploy.ts
import { BaseCommand } from "../../base.js";
import { Args, Flags } from "@oclif/core";
import fs from "node:fs";
import path from "node:path";
import type { ContractData, Hex } from "@nilfoundation/niljs";
import type { Abi } from "abitype";

export default class WalletDeploy extends BaseCommand {
  static override summary = "Deploy a smart contract";
  static override description =
    "Deploy the smart contract with the specified hex-bytecode from stdin or from file";

  static flags = {
    shardId: Flags.integer({
      char: "s",
      description: "Specify the shard ID to interact with",
      required: false,
      default: 1,
    }),
    salt: Flags.integer({
      char: "t",
      description: "The salt for the deploy message",
      required: false,
      default: 0,
    }),
    abiPath: Flags.string({
      char: "a",
      description: "The path to the ABI file",
      required: false,
    }),
    amount: Flags.integer({
      char: "m",
      description: "The amount of default tokens to send",
      required: false,
    }),
    currency: Flags.integer({
      char: "c",
      description:
        'The amount of contract currency to generate. This operation cannot be performed when the "no-wait" flag is set',
      required: false,
      dependsOn: ["currencyName"],
    }),
    currencyName: Flags.string({
      char: "C",
      description: "The name of the currency to generate, required when the currency flag is set",
      required: false,
      dependsOn: ["currency"],
    }),
    noWait: Flags.boolean({
      char: "n",
      description: "Define whether the command should wait for the receipt",
      required: false,
      default: false,
    }),
    compileInput: Flags.string({
      char: "i",
      description:
        "The path to the JSON file with the compilation input. Contract will be compiled and deployed on the blockchain and the Cometa service",
      required: false,
    }),
  };

  static args = {
    filename: Args.string({
      name: "filename",
      required: false,
      description: "The path to the bytecode file",
    }),
    args: Args.string({
      name: "args",
      required: false,
      description: "Constructor arguments for the contract",
      multiple: true,
    }),
  };

  static override examples = ["<%= config.bin %> <%= command.id %>"];

  public async run(): Promise<Hex> {
    const { flags, args } = await this.parse(WalletDeploy);

    if (flags.noWait) {
      if (flags.currency) {
        this.error("The currency flag cannot be set when the no-wait flag is set");
      }
      if (flags.compileInput) {
        this.error("The compileInput flag cannot be set when the no-wait flag is set");
      }
    }

    const { wallet } = await this.setupWallet();

    let abi: Abi | undefined;

    if (flags.abiPath) {
      const abiPath = flags.abiPath;
      const abiFullPath = path.resolve(abiPath);
      const abiFileContent = fs.readFileSync(abiFullPath, "utf8");
      abi = JSON.parse(abiFileContent);
    }

    let bytecode: Hex | Uint8Array;
    let contractData = {} as ContractData;

    if (flags.compileInput) {
      const cometaClient = this.cometaClient ?? this.error("Cometa client is not initialized");
      const compileInputPath = path.resolve(flags.compileInput);
      const compileInputContent = fs.readFileSync(compileInputPath, "utf8");
      contractData = await cometaClient.compileContract(compileInputContent);
      bytecode = contractData.code;
      abi = contractData.abi as unknown as Abi;
    } else {
      const filename = args.filename;
      if (!filename) {
        this.error("at least one arg is required (the path to the bytecode file");
      }
      const fullPath = path.resolve(filename);
      bytecode = fs.readFileSync(fullPath, "utf8") as Hex;
    }

    const { hash, address } = await wallet.deployContract({
      shardId: flags.shardId,
      bytecode: bytecode,
      abi: abi,
      args: args.args?.split(" ") ?? [],
      salt: BigInt(flags.salt),
      value: BigInt(flags.amount ?? 0),
    });

    if (flags.quiet) {
      this.log(address);
    } else {
      this.log("Contract address: ", address);
    }

    if (flags.noWait) {
      return address;
    }

    this.info("Waiting for the contract to be deployed...");
    await this.waitOnTx(hash);
    this.info("Contract successfully deployed");

    if (flags.compileInput) {
      const cometaClient = this.cometaClient ?? this.error("Cometa client is not initialized");
      await cometaClient.registerContractData(contractData, address);
    }

    if (flags.currency) {
      const name =
        flags.currencyName ?? this.error("Currency name is required when the currency flag is set");

      let hash = await wallet.setCurrencyName(name);
      this.info("Waiting for the currency name to be set...");
      await this.waitOnTx(hash);
      this.info("Currency name successfully set");

      hash = await wallet.mintCurrency(BigInt(flags.currency));
      this.info("Waiting for the currency to be minted...");
      await this.waitOnTx(hash);
      this.info("Currency successfully minted");
    }

    return address;
  }
}
